### ISSUE TITLE
High memory usage, killed by OOM daemon

### ISSUE THREAD

**Issue by nerdv2** [xpipe-io/xpipe#400]
Hi, thank you for this project and I have been using it for almost a year now.

These past few months I keep encountering crashes and really sluggish performance and experience with the app due to excessive memory usage xpiped taken. Startup time could take around 1-2 minutes, using around 2-3GB of RAM and keep climbing until I close the application or killed by systemd-oomd.

I am using Git sync without encryption and currently have 12 groups and 384 connections.

Version: Community Edition 13.4.3
OS: Ubuntu 24.04 LTS

![image](https://github.com/user-attachments/assets/cae7b00a-2559-414f-b5e4-0fb4cedf181a)

**Comment by crschnick** [comment]
That must be a bug, anything over 1GB is too much. I have 500+ with git for testing and it doesn't even reach 1GB.

Can you try to create a heap dump with something like https://visualvm.github.io/download.html ?

**Comment by nerdv2** [comment]
Thank you for the quick response @crschnick 

What version of Java do I need to create a heap dump? It seems I can't click the button to create one.
And it seems the heap size is way bigger than it currently consumes, around 1GB, as you said.

![Screenshot from 2024-12-16 16-11-33](https://github.com/user-attachments/assets/1a320ddb-6803-4e74-81cd-cb47f3b21fc7)

**Comment by crschnick** [comment]
Ehm can you use `/opt/xpipe/app/scripts/xpiped_debug.sh` to run it? If that also doesn't work, can you edit the debug script and remove the `-XX:+DisableAttachMechanism` option?

**Comment by crschnick** [comment]
And if you succeed, you can upload the file somewhere and send it to me via email at crschnick@xpipe.io

**Comment by nerdv2** [comment]
Unfortunately running the debug option still doesn't show the option in VisualVM, and there's no disable attach mechanism within the debug script.

![image](https://github.com/user-attachments/assets/45fbed96-a29c-45ea-96bc-3a4be3dc2f1a)

**Comment by crschnick** [comment]
So I will try to release the first XPipe 14 test build in a few days, and I will probably just build a heap dump functionality into it. I think it's best if I come back to you in a couple of days when this is ready

**Comment by crschnick** [comment]
Can you try this: https://github.com/xpipe-io/xpipe-ptb

In the troubleshoot options, there is a heap dump button.

**Comment by nerdv2** [comment]
@crschnick 
Sorry for the late response, catching up tasks before the holidays.

The good news is, there's no memory issue with the preview build but all of the connections is incorrectly imported only showing the group and connection name.

![Screenshot from 2024-12-23 15-51-53-fotor-20241223155433](https://github.com/user-attachments/assets/0b21e616-4057-44bf-a98f-a3ad09d7b2fe)

**Comment by crschnick** [comment]
Hmm, I might have messed up some schema migration there. Let me take a look

**Comment by crschnick** [comment]
Can you try the latest PTB build? That should be fixed

Also, for screenshots, there is now a censor mode in the appearance settings that will black out everything automatically.

**Comment by nerdv2** [comment]
The latest build is working, and initial memory load is good.

I see it still spike quite high, and I observe around 2-3gb of ram usage but recovering to around 1gb after 10 minute of launch which much better than what the production release do.

Do you still want the heap dump file? I could still send it to you.

![image](https://github.com/user-attachments/assets/7a0b5551-02dc-42aa-a8d4-0705bd25bca2)

**Comment by crschnick** [comment]
Yes, you can send me the heap dump file somehow privately via mail (I guess you need to upload it somewhere). Ideally when it is the largest. You should be able to create a heap dump at any time in the troubleshooting settings in xpipe.

**Comment by crschnick** [comment]
I think I was able to fix the issue.

You can try whether the latest release of https://github.com/xpipe-io/xpipe-ptb works for you

**Comment by nerdv2** [comment]
@crschnick 
Whoa, I tried the latest build and the memory usage is really good.

I tried stress testing it, running the app for around an hour with 10 concurrent SSH session while downloading 2gb of archive file and the memory usage only around 200-300mb.

![image](https://github.com/user-attachments/assets/c8388307-614f-4079-9dc4-a03b39a2e2c2)

I think it's probably safe to close this issues after version 14 releases to the public, can't wait and thank you so much for this update.

**Comment by crschnick** [comment]
This is now released in v14
